commands2/command.py

```diff
@@ -0,0 +1,61 @@
+from abc import abstractmethod
+from typing import Generator, Callable, Set, Optional
+
+from .requirement import Requirement, InterruptBehavior
+
+
+class Command:
+    requirements: Set[Requirement] = {}
+    runs_when_disabled: bool = False
+    interruption_behavior: InterruptBehavior = InterruptBehavior.CANCEL_SELF
+
+    @abstractmethod
+    def initialize(self) -> Optional[Generator[None, None, None]]:
+        pass
+
+    def end(self, interrupted: bool):
+        pass
+
+    @property
+    def is_composed(self) -> bool:
+        return getattr(self, "_composed", False)
+
+    def mark_composed(self):
+        setattr(self, "_composed", True)
+
+
+def commandify(
```

Comment:
I think we could support `end` by throwing an exception into the coroutine

Reply:
Yeah I think that would be better. There could be a CommandInterrupted exception.

And it could be used as a try finally without except in most cases.

Reply:
I think that would complicate user-side code.
Honestly, I kindof like the `.end` decorator idea.

Reply:
Well I think ideally most users don't use end at all and this is just a workaround for the few that do. By making it an exception, you have access to the existing context of the coroutine.